src/perf/bin/drvmain.cpp

```diff
@@ -0,0 +1,563 @@
+    WdfDeviceInitSetFileObjectConfig(
+        DeviceInit,
+        &FileConfig,
+        &Attribs);
+    WDF_OBJECT_ATTRIBUTES_INIT_CONTEXT_TYPE(&Attribs, QUIC_DEVICE_EXTENSION);
+
+    Status =
+        WdfDeviceCreate(
+            &DeviceInit,
+            &Attribs,
+            &Device);
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfDeviceCreate failed");
+        goto Error;
+    }
+
+    DeviceContext = QuicPerfCtlGetDeviceContext(Device);
+    RtlZeroMemory(DeviceContext, sizeof(QUIC_DEVICE_EXTENSION));
+    ExInitializePushLock(&DeviceContext->Lock);
+    InitializeListHead(&DeviceContext->ClientList);
+
+    Status = WdfDeviceCreateSymbolicLink(Device, &QuicPerfCtlDeviceSymLink);
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfDeviceCreateSymbolicLink failed");
+        goto Error;
+    }
+
+    WDF_IO_QUEUE_CONFIG_INIT_DEFAULT_QUEUE(&QueueConfig, WdfIoQueueDispatchParallel);
+    QueueConfig.EvtIoDeviceControl = QuicPerfCtlEvtIoDeviceControl;
+    QueueConfig.EvtIoCanceledOnQueue = QuicPerfCtlEvtIoCanceled;
+
+    __analysis_assume(QueueConfig.EvtIoStop != 0);
+    Status =
+        WdfIoQueueCreate(
+            Device,
+            &QueueConfig,
+            WDF_NO_OBJECT_ATTRIBUTES,
+            &Queue);
+    __analysis_assume(QueueConfig.EvtIoStop == 0);
+
+    if (!NT_SUCCESS(Status)) {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            Status,
+            "WdfIoQueueCreate failed");
+        goto Error;
+    }
+
+    QuicPerfCtlDevice = Device;
+    QuicPerfCtlExtension = DeviceContext;
+
+    WdfControlFinishInitializing(Device);
+
+    QuicTraceLogVerbose(
+        PerfControlInitialized,
+        "[perf] Control interface initialized");
+
+Error:
+
+    if (DeviceInit) {
+        WdfDeviceInitFree(DeviceInit);
+    }
+
+    return Status;
+}
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicPerfCtlUninitialize(
+        void
+    )
+{
+    QuicTraceLogVerbose(
+        PerfControlUninitializing,
+        "[perf] Control interface uninitializing");
+
+    if (QuicPerfCtlDevice != nullptr) {
+        NT_ASSERT(QuicPerfCtlExtension != nullptr);
+        QuicPerfCtlExtension = nullptr;
+
+        WdfObjectDelete(QuicPerfCtlDevice);
+        QuicPerfCtlDevice = nullptr;
+    }
+
+    QuicTraceLogVerbose(
+        PerfControlUninitialized,
+        "[perf] Control interface uninitialized");
+}
+
+PAGEDX
+_Use_decl_annotations_
+void
+QuicPerfCtlEvtFileCreate(
+    _In_ WDFDEVICE /* Device */,
+    _In_ WDFREQUEST Request,
+    _In_ WDFFILEOBJECT FileObject
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    PAGED_CODE();
+
+    KeEnterGuardedRegion();
+    ExfAcquirePushLockExclusive(&QuicPerfCtlExtension->Lock);
+
+    do
+    {
```

Comment:
```suggestion
    do {
```

Reply:
We'll need to fix this in the test code then at some point too.